### PR TITLE
Add caret and vcaret

### PR DESCRIPTION
**Pull request by benkiel**
As per discussion here https://github.com/unified-font-object/ufo-spec/issues/89

### CODE REVIEW

File: versions/ufo3/glyphs/glif.md [hunk 1]
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
 
 #### This element has no child elements.
 
+#### Anchor naming conventions
+
+The following are agreed upon standard naming conventions for the `name` of anchors for special uses.
+
+##### hcaret and vcaret
+
+Anchors with a `name` starting with `hcaret` or `vcaret` indicate horizontal or vertical position for the [GDEF Ligature Caret List table]. Anchors with the name `hcaret` provide the `x` value for a ligature caret, and the `y` value is ignored. Likewise, anchors with the name `ycaret` provide the `y` values for a ligature caret, and the `x` value is ignored.
```
Comment (justvanrossum) [inline versions/ufo3/glyphs/glif.md]: I would like the coupling to OpenType GDEF to be looser. In general UFO is export-format agnostic. Maybe mention GDEF caret stuff as an "for example".

Something like
"Anchors with a `name` starting with `hcaret` or `vcaret` indicate horizontal or vertical caret positions for ligature glyphs, for example to compile the OpenType GDEF Ligature Caret List table."

File: versions/ufo3/glyphs/glif.md [hunk 2]
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
...
+
+Both names may be followed by an underscore and arbitrary text (`_X`) for designers to have unique caret position names. The order of carets written to the GDEF table is based upon ascending value order, the `_X` text is not taken into account.
+
+`hcaret` and `vcaret` must not both occur in a glyph.
+
```
Comment (justvanrossum) [inline versions/ufo3/glyphs/glif.md:183]: Maybe mention that it is expected to have N-1 caret positions for N ligature components.

**Comment by moyogo** [comment]
I now think the distinction between horizontal and vertical carets is not necessary. The glyphs direction should indicate what type of caret it is. It can be just named `caret` instead of `hcaret` or `vcaret`.

**Comment by anthrotype** [comment]
in theory (though unlikely) the same glyph could potentially be mapped to different unicodes whose script have conflicting directions.... but we probably should not support this limit case.

**Comment by benkiel** [comment]
@moyogo would it be helpful to have the distinction for cases where the direction isn't clear? I know the h/v is fussier but it is explicitly clear.

**Comment by justvanrossum** [comment]
I would prefer to be explicit here instead of depending on the compiler having to determine verticality based on the unicode. May save us some epic explanations, though :)

**Comment by anthrotype** [comment]
have we thought about using guidelines instead of anchors? I believe guidelines can have a name assigned. We just require that a glyph only has either all vertical or all horizontal "caret[_X]?" guidlines. Haven't fully thought this through.

**Comment by typoman** [comment]
Guidelines make more sense as it's also visually easier to see the caret in the glyph view too.

**Comment by benkiel** [comment]
@anthrotype the one wrinkle with guidelines would be that there are glyph and font/layer level guidelines. That's just saying "only glyph level guidelines", but I don't know if that would trip things up in Glyphs or not

**Comment by anthrotype** [comment]
Glyphs has its own way to specify and generate this GDEF ligature carets. Even the current proposal would need some work on their side in order to support it. They use the name "caret_N" where N is a digit starting from 1. I don't think they accept just caret, let alone hcaret.
Guidelines seem to me more natural for displaying a caret position, but I don't have a strong opinion on that.

**Comment by benkiel** [comment]
Based on that, that's an argument for guideline and just `caret[_X]`. I don't feel strongly that this needs to be as anchors, @justvanrossum?

**Comment by justvanrossum** [comment]
Guidelines are drawing aids, and I don't thing they've ever been used to declare data in this way. Visually, I understand the point, but structurally in UFO it feels wrong. But that's about as solid as my argument gets I'm afraid.

**Comment by justvanrossum** [comment]
> Glyphs has its own way to specify and generate this GDEF ligature carets. Even the current proposal would need some work on their side in order to support it. They use the name "caret_N" where N is a digit starting from 1. I don't think they accept just caret, let alone hcaret.

Do they use anchors this way? And do they support vertical ligature caret positions? If so, how? Unicode magic?

**Comment by anthrotype** [comment]
> do they support vertical ligature caret positions?

I don't think so. @schriftgestalt?

**Comment by justvanrossum** [comment]
I can also see us opting for `caret_N` and `vcaret_N` if that will result in cheap interoperability with Glyphs for horizontal carets.

**Comment by schriftgestalt** [comment]
>> do they support vertical ligature caret positions?

>I don't think so. @schriftgestalt?

Does anyone has a sample font?

**Comment by justvanrossum** [comment]
@schriftgestalt just to make sure: does Glyphs indeed use anchors named `caret_1`, `caret_2` etc. to compile GDEF Ligature Caret Lists?

And if so: is the number significant? Is it an error to have them in arbitrary order? For example from left to right: `caret_2`, `caret_1`, `caret_3`?

**Comment by schriftgestalt** [comment]
The number doesn't seem do be significant. Glyphs writes the number sorted by anchor name but makeOTF is sorting the values.

**Comment by justvanrossum** [comment]
One more thing, can you confirm whether the N in `caret_N` even needs to be a number? I see that glyphsLib only looks for `caret_` and ignores everything that comes after.

File: versions/ufo3/glyphs/glif.md [hunk 3]
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
 
 #### This element has no child elements.
 
+#### Anchor naming conventions
+
+The following are agreed upon standard naming conventions for the `name` of anchors for special uses.
+
+##### Ligature hcaret and vcaret
+
+Anchors with a `name` starting with `hcaret` or `vcaret` indicate horizontal or vertical position for ligature glyphs, for example to complie the [OpenType GDEF Ligature Caret List table]. Anchors with the name `hcaret` provide the `x` value for a ligature caret, and the `y` value is ignored. Likewise, anchors with the name `ycaret` provide the `y` values for a ligature caret, and the `x` value is ignored.
```
Comment (justvanrossum) [inline versions/ufo3/glyphs/glif.md]: - As per the discussion, I would love for `hcaret` to be renamed to `caret`, and that we require the anchor name to start with `caret_` and `vcaret_` respectively, in both occurances.
- "complie" -> "compile"
- "ycaret" -> "vcaret"

File: versions/ufo3/glyphs/glif.md [hunk 4]
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
...
+
+Both names may be followed by an underscore and arbitrary text (`_X`) for designers or applications to have unique caret position names.
+
+`hcaret` and `vcaret` must not both occur in a glyph. There must be at least `n`-1 caret postions for `n` ligature components.
```
Comment (justvanrossum) [inline versions/ufo3/glyphs/glif.md]: - hcaret -> caret
- also add underscores here
- "at least" -> "exactly"
- minor: I would perhaps not use backticks around "n" here, but simply use unformatted cap N: "There must be exactly N-1 caret postions for N ligature components."

Comment (benkiel) [reply versions/ufo3/glyphs/glif.md]: What if someone wanted to skip over a glyph in a ligature? I don't think we should say exactly, but "no more than".

**Comment by benkiel** [comment]
I took out the requirement that only `caret` or `vcaret` be in a glyph, as it's an arbitrary limit that doesn't need to be there.

**Comment by schriftgestalt** [comment]
> One more thing, can you confirm whether the N in caret_N even needs to be a number? I see that glyphsLib only looks for caret_ and ignores everything that comes after.

The number is ignored so anything is fine.

File: versions/ufo3/glyphs/glif.md [hunk 5]
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
 
 #### This element has no child elements.
 
+#### Anchor naming conventions
+
+The following are agreed upon standard naming conventions for the `name` of anchors for special uses.
+
+##### Ligature Carets
+
+Anchors with a `name` starting with `caret_` or `vcaret_` indicate horizontal or vertical position for ligature glyphs, for example to compile the [OpenType GDEF Ligature Caret List table]. Anchors with the name `caret_` provide the `x` value for a ligature caret, and the `y` value is ignored. Likewise, anchors with the name `vcaret_` provide the `y` values for a ligature caret, and the `x` value is ignored.
```
Comment (justvanrossum) [inline versions/ufo3/glyphs/glif.md]: "indicate horizontal or vertical position for ligature glyphs" ->
"indicate horizontal or vertical caret positions for ligature glyphs"

File: versions/ufo3/glyphs/glif.md [hunk 6]
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
...
+
+Both names may be followed arbitrary text (`caret_1`, `caret_blah`, etc) for designers or applications to have unique caret position names.
+
+There must be no more than N-1 caret postions for N ligature components.
```
Comment (justvanrossum) [inline versions/ufo3/glyphs/glif.md]: I would still say "There must be exactly N-1 caret positions for N ligature components." (also note the typo in positions)

File: versions/ufo3/glyphs/glif.md [hunk 7]
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
 
 #### This element has no child elements.
 
+#### Anchor naming conventions
+
+The following are agreed upon standard naming conventions for the `name` of anchors for special uses.
+
+##### Ligature Carets
+
+Anchors with a `name` starting with `caret_` or `vcaret_` indicate horizontal or vertical caret positions for ligature glyphs, for example to compile the [OpenType GDEF Ligature Caret List table]. Anchors with the name `caret_` provide the `x` value for a ligature caret, and the `y` value is ignored. Likewise, anchors with the name `vcaret_` provide the `y` values for a ligature caret, and the `x` value is ignored.
```
Comment (justvanrossum) [inline versions/ufo3/glyphs/glif.md]: Sorry, a few more nitpicky things:
"Anchors with the name `caret_`" -> "Anchors with a name starting with `caret_`"
"value for a ligature caret" -> "value for a ligature caret position"
"anchors with the name `vcaret_`" -> "anchors with a name starting with `vcaret_`"
"provide the `y` values for a ligature caret" -> "provide the `y` value for a ligature caret position"

**Comment by benkiel** [comment]
My bad. Fixed the plural anchors.

**Review by justvanrossum — APPROVED** [review]
Looks good to me. Please request a review from Tal.

**Comment by benkiel** [comment]
@typesupply Can you review and approve?

**Comment by typesupply** [comment]
Looks good to me.

**Comment by benkiel** [comment]
@typesupply Thank you! Can you merge go ahead and merge, I don't have those rights.

**Comment by khaledhosny** [comment]
An idea for future UFO versions; either specify that anchors can have x or y omitted and then they would be displayed as lines instead of points, or define a new structure that would have such property. This can be used for ligature caret as well as MATH italic correction and top accent positions (both are x only), and any potential new property that would need this.

**Pull request merged**